### PR TITLE
database/sinkdb: introduce IfMatch condition

### PR DESCRIPTION
**Pull request by kr**
This lets clients do safe conditional writes.

Also update net/http/authz to use the new feature.

### CODE REVIEW

**Comment by kr** [comment]
PTAL

File: database/sinkdb/op.go [hunk 1]
```diff
@@ -49,6 +49,22 @@ func IfNotExists(key string) Op {
 	}
 }
 
+// IfMatch encodes a conditional to make an instruction
+// successful only if the stored version matches v.
+// It checks the key returned by method Key on v.
```
Comment (kr) [inline database/sinkdb/op.go]: This is awkward wording.

Comment (kr) [inline database/sinkdb/op.go]:
```
IfMatch encodes a conditional to make an instruction
successful only if v's key's stored version matches v.
```

Comment (kr) [inline database/sinkdb/op.go]:
```
IfMatch encodes a conditional to make an instruction
successful only if the version stored in v's key matches v.
```

File: database/sinkdb/op.go [hunk 2]
```diff
@@ -49,6 +49,22 @@ func IfNotExists(key string) Op {
...
+//
+// If v.Exists() is false,
+// IfMatch(v) is equivalent to IfNotExists(v.Key()).
+func IfMatch(v Version) Op {
```
Comment (jbowens) [inline database/sinkdb/op.go]: 🚲 🏠  any thoughts on `IfUnmodified`?

Comment (kr) [inline database/sinkdb/op.go]: `IfUnmodified` is fine by me.

Comment (kr) [inline database/sinkdb/op.go]: or `IfNotModified` to match `IfNotExists`?

**Comment by jbowens** [comment]
lgtm, tho `IfMatch` feels a little ambiguous

File: database/sinkdb/op.go (hunk 2 above)
Comment (kr) [inline database/sinkdb/op.go]: We could hypothetically also have `IfExists` and `IfModified` though those ones seem much less useful.

Comment (kr) [inline database/sinkdb/op.go]: (I started with IfMatch because that's what HTTP calls the equivalent header field, but that's not a very good reason.)

**Pull request merged**